Forecasting models (#163)
nfidd/nfidd: sessions/forecasting-models.qmd

If not using the Visual Editor please remember that the code in the session needs to be run inside the course repository so that the `here()` commands below find the stan model files.
:::

# An overview of forecasting models

> See the slides for an overview of forecasting models.

Reviewer: add paragraph in new issue (I volunteer)

# What did we learn about the random walk model from the [forecasting concepts session](forecasting-concepts)?

> - It was unbiased when the reproduction number was constant but when the reproduction number was reducing due to susceptible depletion, the random walk model overestimated the number of cases systematically.

Reviewer: was it really ever constant? I would have thought than on the log scale there is a similar amount of overestimation anywhere given more or less linear decline of R.

- It did relatively well at short-term forecasting indicating that it was capturing some of the underlying dynamics of the data generating process.

::: {.callout-note, collapse = TRUE}

> ## What did a geometric random walk model look like again?

Reviewer: to write function where geometric random walk is first introduced #122

n <- 100
R <- rep(NA, n)
R[1] <- rnorm(1, 1, 0.1)
for (i in 2:n) {
  R[i] <- R[i-1] + rnorm(1, 0, 0.1)
}

data <- tibble(t = 1:n, R = exp(R))

ggplot(data, aes(x = t, y = R)) +
  geom_line() +
  labs(title = "Simulated data from a random walk model",
       x = "Time",
       y = "R")
```

:::

# Forecasting models as a spectrum. 

A large part of this course has been about showing the importance of understanding and modelling the underlying mechanisms of the data generating process.
However, in many forecasting scenarios, we may not have a good understanding of the underlying mechanisms, or the data may be too noisy to make accurate predictions.
The worst case is that we have mistaken beliefs about the underlying mechanisms and use these to make predictions which are systematically wrong and misleading.
In these cases, forecasters of often use _statistical models_ to make predictions which have little or no mechanistic basis.
In our work, we have found that a combination of mechanistic and statistical models can be very powerful but that identifying the best model for a given forecasting task can be challenging.

## Adding more mechanistic structure to the renewal model

> One way to improve the renewal model is to add more mechanistic structure. In the [forecasting concepts session](forecasting-concepts), we saw that the renewal model was unbiased when the reproduction number was constant but that it overestimated the number of cases when the reproduction number was reducing due to susceptible depletion.

Reviewer: ```suggestion
One way to improve the renewal model is to add more mechanistic structure. In the [forecasting concepts session](forecasting-concepts), we saw that the renewal model was making unbiased forecasts when the reproduction number was constant but that it overestimated the number of cases when the reproduction number was reducing due to susceptible depletion.
```

$$
R_t = \frac{S_{t-1}}{N} R_0
$$

This approximates susceptible depletion as a linear function of the number of susceptibles in the population. This is a simplification but it is a good starting point.

::: {.callout-note, collapse = TRUE}
## What behaviour would we expect from this model?

```{r}
set.seed(123)
n <- 100
N <- 1000
R0 <- 1.5
S <- rep(NA, n)
S[1] <- N
R <- rep(NA, n)
R[1] <- R0
I <- rep(NA, n)
I[1] <- 1
for (i in 2:n) {
  R[i] <- (S[i-1]) / N * R0
  I[i] <- I[i-1] * R[i]
  S[i] <- S[i-1] - I[i]
}

data <- tibble(t = 1:n, R = R)

ggplot(data, aes(x = t, y = R)) +
  geom_line() +
  labs(title = "Simulated data from an SIR model",
       x = "Time",
       y = "R")
```

The key assumptions we are making here are: 

- The population is constant and we rouchgly know the size of the population.

> - The reproduction number is constant (aside from the effect of susceptible depletion)

Reviewer: ```suggestion
- The reproduction number only changes due to susceptible depletion
```

> - The number of cases is proportional to the number of susceptibles in the population.

Reviewer: ```suggestion
- The number of new cases at each time is proportional to the number of susceptibles in the population.
```

We've coded this up as a stan model in `stan/mechanistic-r.stan`. See `stan/functions/pop_bounded_renewal.stan` for the function which calculates the reproduction number. Lets load the model:

```{r}

> mech_mod <- cmdstan_model(here("stan/mechanistic-r.stan"))

Reviewer: ```suggestion
mech_mod <- cmdstan_model(here("stan", "mechanistic-r.stan"))
```
does your version work on windows (which uses backslash)?

mech_mod$print(line_numbers = TRUE)
```

## Adding more statistical structure to the renewal model

> Rather than adding more mechanistic structure to the renewal model, we can add more statistical structure.

Reviewer: ```suggestion
Adding more mechanistic structure is not always possible and, if we don't specify mechanisms correctly, might make forecasts worse.
Rather than adding more mechanistic structure to the renewal model, we could add more statistical structure.
```

Before we do this, we need to think about what we want from a forecasting model.
As we identified above, we want a model which is unbiased and which has good short-term forecasting properties.
We know that we want it to be able to adapt to trends in the reproduction number and that we want it to be able to capture the noise in the data.

> A statistical term that can be used to describe capturing a trend is saying that the time series is _non-stationary_.

Reviewer: ```suggestion
A statistical term that can be used to describe capturing a trend is saying that the time series is _non-stationary_.
More specifically, a _stationary_ time series is defined as one that looks indistinguishable at all times.
In infectious disease epidemiology, this would only be expected for endemic diseases without external seasonal influence.
```

The random walk model we used in the [forecasting concept](forecasting-concepts) session is a special case of a more general class of models called _autoregressive (AR) models_. 
AR models are a class of models which predict the next value in a time series as a linear combination of the previous values in the time series.

> The random walk model is specfically a special case of an AR(1) model where the next value in the time series is predicted as the previous value, multipled by a value between 1 and -1 , plus some noise. For the log-transformed reproduction number ($ log(R_t) $), the model is:

Reviewer: ```suggestion
The random walk model is specifically a special case of an AR(1) model where the next value in the time series is predicted as the previous value, multiplied by a value between 1 and -1 , plus some noise. For the log-transformed reproduction number ($ log(R_t) $), the model is:
```

$$
log(R_t) = \phi log(R_{t-1}) + \epsilon_t
$$

where $\epsilon_t$ is a normally distributed error term with mean 0 and standard deviation $\sigma$ and $\phi$ is a parameter between -1 and 1. If we restict $\phi$ to be between 0 and 1, we get a model which is biased towards a reproduction number of 1 but which can still capture trends in the data that decay over time.

::: {.callout-note, collapse = TRUE
## What behaviour would we expect from this model?

```{r}
set.seed(123)
n <- 100
phi <- 0.4
sigma <- 0.1
log_R <- rep(NA, n)
log_R[1] <- rnorm(1, 0, sigma)
for (i in 2:n) {
  log_R[i] <- phi * log_R[i-1] + rnorm(1, 0, sigma)
}
data <- tibble(t = 1:n, R = exp(log_R))

ggplot(data, aes(x = t, y = R)) +
  geom_line() +
  labs(title = "Simulated data from an exponentiated AR(1) model",
       x = "Time",
       y = "R")
```
:::

> However, we probably don't want a model which is biased towards a reproduction number of 1 (though as we know the future in this case it actually could be a reasonable choice!). So what should we do?

Reviewer: ```suggestion
However, we probably don't want a model which is biased towards a reproduction number of 1 (unless we have good reason to believe that is the expected behaviour). So what should we do?
```

> Returning to the idea that the reproduction number is a _non-stationary_ time series, we can use a method from the field of time series analysis called _differencing_ to make the time series stationary. This involves taking the difference between consecutive values in the time series. For the log-transformed reproduction number, this would be:

Reviewer: needs motivation for why we want to make it stationary?

)
```
:::
:::

# Evaluating forecasts from our models

As in the [forecasting concepts session](forecasting-concepts), we have fit these models to a range of forecast dates so you don't have to wait for the models to fit. We will now evaluate the forecasts from the mechanistic and statistical models.

```{r load_forecasts, eval = FALSE}
forecasts <- bind_rows(
  rw_forecasts |>
    mutate(model = "random_walk"),
  mech_forecasts |>
    mutate(model = "mechanistic"),
  stat_forecasts |>
    mutate(model = "statistical")
)
```

::: {.callout-tip}
We generated these forecasts using the code in `data-raw/generate-example-forecasts.r` which uses the same approach we just took for a single forecast date but generalises it to many forecasts dates.

Some important things to note about these forecasts:

  - We used a 14 day forecast horizon.
  - Each forecast used all the data up to the forecast date.
  - We generated 1000 posterior samples for each forecast.
  - We started forecasting 3 weeks into the outbreak and then forecast every 7 days until the end of  the data (excluding the last 14 days to allow a full forecast).
  - We used pregenerated simulated data:

```{r}

> example_onset_df

Reviewer: is this different from `onset_df` above?
